airflow/jobs/scheduler_job.py

```diff
@@ -807,6 +810,37 @@ def _clean_tis_without_dagrun(self, session):
                     raise
             guard.commit()
 
+    @provide_session
+    def _clean_tis_without_dag(self, session: Session = None):
+        """Fails task instances and DagRuns of DAGs that no longer exist in the dagbag/DB"""
+        states_to_check = State.unfinished - frozenset([State.NONE, State.SHUTDOWN])
+        tis = session.query(TI).filter(TI.state.in_(states_to_check)).all()
+        missing_dags = {}
+        dag_runs = set()
+
+        for ti in tis:
+            if ti.dag_id in missing_dags:
+                ti.set_state(State.FAILED, session=session)
+                continue
+            # Dag no longer in dagbag?
+            if not self.dagbag.has_dag(ti.dag_id, session=session):
+                ti.set_state(State.FAILED, session=session)
```

Comment:
I wonder if this (and L823) should be set to State.REMOVED? It might be clearer for debugging for the user than a failure without any logs.